In [14]:
#Import Libraries
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
plt.style.use('seaborn')
import seaborn as sns



In [28]:
#Import dataset
df = pd.read_csv('listings.csv')

#Checking Shape
print(f'{df.shape[0]} examples, {df.shape[1]} features')

# ... and duplicates
print("It contains {} duplicates.".format(df.duplicated().sum()))


24551 examples, 106 features
It contains 0 duplicates.


In [34]:
df.head()

id                        listing_url       scrape_id last_scraped  \
0  1944  https://www.airbnb.com/rooms/1944  20190812143128   2019-08-13   
1  2015  https://www.airbnb.com/rooms/2015  20190812143128   2019-08-13   
2  3176  https://www.airbnb.com/rooms/3176  20190812143128   2019-08-13   
3  3309  https://www.airbnb.com/rooms/3309  20190812143128   2019-08-13   
4  6883  https://www.airbnb.com/rooms/6883  20190812143128   2019-08-13   

                                                name  \
0     cafeheaven Pberg/Mitte/Wed for the summer 2019   
1   Berlin-Mitte Value! Quiet courtyard/very central   
2                    Fabulous Flat in great Location   
3                  BerlinSpot Schöneberg near KaDeWe   
4  Stylish East Side Loft in Center with AC & 2 b...   

                                             summary  \
0  Private, bright and friendly room. You'd be sh...   
1  Great location!  30 of 75 sq meters. This wood...   
2  This beautiful first floor apartment  is situa...   
3  First of all: I prefer short-notice bookings. ...   
4                                                NaN   

                                               space  \
0  The room is very large, private, cozy, bright,...   
1  A+++ location! This „Einliegerwohnung“ is an e...   
2  1st floor (68m2) apartment on Kollwitzplatz/ P...   
3  Your room is really big and has 26 sqm, is ver...   
4  Stay in a stylish loft on the second floor and...   

                                         description experiences_offered  \
0  Private, bright and friendly room. You'd be sh...                none   
1  Great location!  30 of 75 sq meters. This wood...                none   
2  This beautiful first floor apartment  is situa...                none   
3  First of all: I prefer short-notice bookings. ...                none   
4  Stay in a stylish loft on the second floor and...                none   

                               neighborhood_overview  \
0  near all the trendy cafés and flea markets and...   
1  It is located in the former East Berlin area o...   
2  The neighbourhood is famous for its variety of...   
3  My flat is in the middle of West-Berlin, direc...   
4  The emerging and upcoming East of the new hip ...   

                                               notes  \
0  Please check-in by 10pm, and we can't accept c...   
1  This is my home, not a hotel. I rent out occas...   
2  We welcome FAMILIES and cater especially for y...   
3  The flat is a strictly non-smoking facility! A...   
4  Parking in an underground parking lot is avail...   

                                             transit  \
0  There is ample street parking. We are near 2  ...   
1  Close to U-Bahn U8 and U2 (metro), Trams M12, ...   
2  We are 5 min walk  away from the tram M2, whic...   
3  The public transportation is excellent: Severa...   
4  Location: - Very close to Alexanderplatz just ...   

                                              access  \
0  Your room, the bathroom and the kitchen, and o...   
1  Simple kitchen/cooking, refrigerator, microwav...   
2  The apartment will be entirely yours. We are c...   
3  I do have a strictly non-smoker-flat.  Keep th...   
4  More details: - Electricity, heating fees and ...   

                                         interaction  \
0  I'll be traveling a lot in the summer and not ...   
1                                   Always available   
2  Feel free to ask any questions prior to bookin...   
3  I'm working as a freelancing photographer. My ...   
4  I rent out my space when I am travelling so I ...   

                                         house_rules  thumbnail_url  \
0  Please do not use the wireless Internet access...            NaN   
1  No parties No events No pets No smoking, not e...            NaN   
2  It’s a non smoking flat, which likes to be tre...            NaN   
3  House-Rules and Information ..............(deu...            NaN   
4  No Pets.  No loud Parties.  Smoking only on

In [206]:
#Look every column in detail
for column in df.columns:
    print('______________________________')
    print('\nCOLUMN NAME:' ,column)
    print('______________________________')
    temp_column= column
    print('NaN values',df[temp_column].isna().sum())
    print(df[temp_column].value_counts())
    
    

______________________________

COLUMN NAME: id
______________________________
NaN values 0
1944        0.000041
24684132    0.000041
24689832    0.000041
24689619    0.000041
24689441    0.000041
              ...   
13683453    0.000041
13681045    0.000041
13677036    0.000041
13673149    0.000041
37698736    0.000041
Name: id, Length: 24551, dtype: float64
______________________________

COLUMN NAME: listing_url
______________________________
NaN values 0
https://www.airbnb.com/rooms/1944        0.000041
https://www.airbnb.com/rooms/24684132    0.000041
https://www.airbnb.com/rooms/24689832    0.000041
https://www.airbnb.com/rooms/24689619    0.000041
https://www.airbnb.com/rooms/24689441    0.000041
                                           ...   
https://www.airbnb.com/rooms/13683453    0.000041
https://www.airbnb.com/rooms/13681045    0.000041
https://www.airbnb.com/rooms/13677036    0.000041
https://www.airbnb.com/rooms/13673149    0.000041
https://www.airbnb.com/rooms/3769873

NaN values 0
Berlin, Berlin, Germany                                     0.989084
Berlin, Germany                                             0.003666
., Berlin, Germany                                          0.000448
Schöneberg, Berlin, Germany                                 0.000407
Weissensee, Berlin, Germany                                 0.000326
Mitte, Berlin, Germany                                      0.000285
Berlin, Schleswig-Holstein, Germany                         0.000285
Berlin, Mitte, Germany                                      0.000244
*, Berlin, Germany                                          0.000244
Berlin , Berlin, Germany                                    0.000244
Berlin, Zehlendorf, Germany                                 0.000163
Wilmersdorf, Berlin, Germany                                0.000163
Berlin, Brandenburg, Germany                                0.000163
Berlin , Germany                                            0.000163
DE, Germany          

COLUMN NAME: cancellation_policy
______________________________
NaN values 1
flexible                       0.396986
moderate                       0.300815
strict_14_with_grace_period    0.297108
super_strict_30                0.002566
super_strict_60                0.002444
strict                         0.000081
Name: cancellation_policy, dtype: float64
______________________________

COLUMN NAME: require_guest_profile_picture
______________________________
NaN values 0
f    0.993564
t    0.006436
Name: require_guest_profile_picture, dtype: float64
______________________________

COLUMN NAME: require_guest_phone_verification
______________________________
NaN values 0
f    0.989084
t    0.010916
Name: require_guest_phone_verification, dtype: float64
______________________________

COLUMN NAME: calculated_host_listings_count
______________________________
NaN values 0
1     0.748483
2     0.123742
3     0.039347
4     0.018085
6     0.009287
7     0.009124
5     0.008554
9     0.0054

In [208]:
# Deciding which columns to work with

#1 City, street, state are mostly same for all, id is unique and scrape_id is same for all.
#2 experiences_offered, thumbnail_url, medium_url are  empty
#3 picture_url, host_thumbnail_url, host_picture_url links are broken.

columns_to_discard = ['city','id','scrape_id','experiences_offered','thumbnail_url','medium_url',
                     'picture_url','xl_picture_url','host_acceptance_rate','host_thumbnail_url',
                     'host_picture_url','street','state','market','smart_location','country_code',
                     'country','has_availability','calendar_last_scraped','license','jurisdiction_names'
                     'is_business_travel_ready',]

#Columns with working link: Can be used for data scraping
linked_columns= ['host_url']

#Columns that can be analyzed with NLP
text_columns = ['name','summary','space','description','neighborhood_overview',
               'notes','transit','access','interaction','house_rules','host_about']

#Date columns can be feature-engineered
date_columns = ['host_since','first_review','last_review',]

#Categorical columns (will be converted to categories)
cat_columns = ['last_scraped','host_response_time','host_response_rate','host_is_superhost',
              'host_neighbourhood','host_verifications','host_has_profile_pic','host_identity_verified',
              'neighbourhood','neighbourhood_cleansed','neighbourhood_group_cleansed','is_location_exact',
              'property_type','room_type','bed_type','amenities','calendar_updated','requires_license',
              'instant_bookable','cancellation_policy','require_guest_profile_picture',
               'require_guest_phone_verification'
              ]

target_column = 'price'


### Notes:

* Interesting feature is text german or english
* Didn't understand host_listings_count and host_total_listings_count (will look more closely)
They are almost same (only 51 of them different from 24551)
* Zipcode should be investigated more (to understand the correlation between numbers and location
if there is any.)
* How bathrooms can be a decimal number like 2.5 =)
* Square feet has 0 values this should be filled with KNN.
* weekly/monthly prices are mostly NaN
* What is the difference between minimum_nights vs minimum_minimum_nights (23748 of them are same)
* Interesting visualizations for review columns (Average is not 5 for example. So we may shift the average)
* host_neighbourhood can be processed as inside the city vs outside the city. (Need Berlin's neighbourhoods)


NaN values 4328
0.02     616
0.04     555
0.05     504
1.00     479
0.08     450
        ... 
6.87       1
6.13       1
5.32       1
9.10       1
15.00      1
Name: reviews_per_month, Length: 795, dtype: int64


In [167]:
x1 =df.minimum_nights
x2 = df.minimum_maximum_nights

(x1 == x2).sum()

299

In [104]:
x1 = df.host_listings_count
x2 = df.host_total_listings_count

(x1 == x2).sum()

24500

In [36]:
df.calendar_last_scraped.value_counts()

2019-08-12    13117
2019-08-13    11432
2019-09-19        2
Name: calendar_last_scraped, dtype: int64

In [44]:
df.host_id.value_counts()


id  listing_url  scrape_id  last_scraped  name  summary  space  \
host_id                                                                     
2058        1            1          1             1     1        1      1   
2164        1            1          1             1     1        1      1   
2217        6            6          6             6     6        6      6   
3718        1            1          1             1     1        1      1   
4108        1            1          1             1     1        1      1   
...        ..          ...        ...           ...   ...      ...    ...   
284720636   1            1          1             1     1        1      0   
284745099   1            1          1             1     1        1      1   
284786105   1            1          1             1     1        1      0   
284806538   1            1          1             1     1        1      0   
284936716   1            1          1             1     1        1      1   

           description  experiences_offered  neighborhood_overview  notes  \
host_id                                                                     
2058                 1                    1                      1      0   
2164                 1                    1                      1      1   
2217                 6                    6                      6      4   
3718                 1                    1                      1      1   
4108                 1                    1                      1      1   
...                ...                  ...                    ...    ...   
284720636            1                    1                      0      0   
284745099            1                    1                      1      0   
284786105            1                    1                      1      0   
284806538            1                    1                      1      0   
284936716            1                    1                      1      0   

           transit  access  interaction  house_rules  thumbnail_url  \
host_id                                                               
2058             1       0            0            0              0   
2164             1       1            1            1              0   
2217             6       6            6            5              0   
3718             1       1            1            1              0   
4108             1       1            1            1              0   
...            ...     ...          ...          ...            ...   
284720636        1       0            1            0              0   
284745099        1       0            1            0              0   
284786105        1       0            1            1              0   
284806538        1       0            1            0              0   
284936716        0       0            1            1              0   

           medium_url  picture_url  xl_picture_url  host_url  host_name  \
host_id                                                                   
2058                0            1               0         1          1   
2164                0            1               0         1          1   
2217                0            6               0         6          6   
3718                0            1               0         1          1   
4108                0            1               0         1          1   
...               ...          ...             ...       ...        ...   
284720636           0            1               0         1          1   
284745099           0            1               0         1          1   
284786105           0            1               0         1          1   
284806538           0            1               0         1          1   
284936716           0            1               0         1          1   

           host_since  host_location  host_about  host_response_time  \
host_id                            

In [46]:
df.scrape_id.value_counts()

20190812143128    24551
Name: scrape_id, dtype: int64